aws/resource_aws_instance.go

```diff
@@ -1838,6 +1901,19 @@ func buildAwsInstanceOpts(
 		opts.InstanceInitiatedShutdownBehavior = aws.String(v)
 	}
 
+	if v, ok := d.GetOk("metadata_options"); ok {
+		if mo, ok := v.([]interface{})[0].(map[string]interface{}); ok {
+			opts.MetadataOptions = &ec2.InstanceMetadataOptionsRequest{
+				HttpEndpoint: aws.String(mo["http_endpoint"].(string)),
+			}
+			if mo["http_endpoint"].(string) == ec2.InstanceMetadataEndpointStateEnabled {
+				// These parameters are not allowed unless HttpEndpoint is enabled
+				opts.MetadataOptions.HttpTokens = aws.String(mo["http_tokens"].(string))
+				opts.MetadataOptions.HttpPutResponseHopLimit = aws.Int64(int64(mo["http_put_response_hop_limit"].(int)))
+			}
+		}
```

Comment:
Similar to creating a flatten function above, the inverse expand function could be created and used in the multiple places with this logic:

```go
func expandEc2InstanceMetadataOptions(l []interface{}) *ec2.MetadataOptions {
	m, ok := l[0].(map[string]interface{})

	if !ok {
		return nil
	}

	opts := &ec2.InstanceMetadataOptionsRequest{
		HttpEndpoint: aws.String(m["http_endpoint"].(string)),
	}

	if m["http_endpoint"].(string) == ec2.InstanceMetadataEndpointStateEnabled {
		// These parameters are not allowed unless HttpEndpoint is enabled

		if v, ok := m["http_tokens"].(string); ok && v != "" {
			opts.HttpTokens = aws.String(v)
		}

		if v, ok := m["http_put_response_hop_limit"].(int); ok && v != 0 {
			opts.HttpPutResponseHopLimit = aws.Int64(int64(v))
		}
	}

	return opts
}

// ...

MetadataOptions: expandEc2InstanceMetadataOptions(d.Get("metadata_options").([]interface{})),
```